### ISSUE TITLE
Adding of duplicate CIs to an archive item

### ISSUE THREAD

**Issue by HarmvZ** [comic/grand-challenge.org#2397]
The set of CIs for an archive item should be unique. However, it is currently possible to add multiple CIV with the same CI for a single archive item. This is because the check to see if an a CI already exists and should be removed for an archive item is done asynchronously. So while the task to add a CIV to an item is queued, another task can be spawned to add another CIV for the same CI. This will result in both CIVs being added to the item so duplicate CIs on a single item. 

I'm not sure how to mitigate this issue. We could try to check the task queue (if that is even possible) before adding a new task to add a CIV if there are already tasks for that CI in it. Then the new task should be cancelled/not triggered. Because of the asynchronous nature it will then probably be impossible to properly show an error about this to the user. Probably there is a better way, but I'm not seeing it currently.

This issue is further magnified because it is not clear that an upload is processing when adding a new CIV to an archive item. When you add a file type CIV to an archive item it will redirect you back to the archive item list view. When the CIV adding task is still queued after you are redirected which is quite likely, you will not see any indication that it is in the process of being added on that page. Only when you check https://grand-challenge.org/cases/uploads/ you can see the upload session status. So a user could think something went wrong and try to add it again which will result in the issue described above. Maybe it is good to redirect the user to the cases/uploads page after adding a CIV so that they see that the task is queued.

**Comment by MikeOverkamp-diag** [comment]
What I think we should do:

Rather than calling `update_archive_item_update_kwargs` separetly in the [view](https://github.com/comic/grand-challenge.org/blob/main/app/grandchallenge/archives/views.py#L419) and in the [serializer](https://github.com/comic/grand-challenge.org/blob/main/app/grandchallenge/archives/serializers.py#L99), it should be chained in the `start_archive_item_update_tasks` task. I think the issue now arises because the the updating is separate from the initial check, which means they can run in parallel, while there is a dependency and the check should always happen before the actual update. What do you think @jmsmkn ?

**Comment by jmsmkn** [comment]
Don't think that is it, although `update_archive_item_update_kwargs` is in `tasks.py` it is not a Celery task, and is being called synchronously, so will always be run before `start_archive_item_update_tasks` anyway.

**Comment by MikeOverkamp-diag** [comment]
Ah I didn't look properly, sorry about that.

**Comment by MikeOverkamp-diag** [comment]
However, wouldn't making it a task and chaining it in front of the other tasks guarantee that there is no time between the check and the execution of the rest of the code? Or alternatively, we can make the check part of the update task. Now the check gets performed on form save, while the rest of the logic gets added to the queue and might not execute until a lot later.

**Comment by jmsmkn** [comment]
Chaining just creates a list of tasks, with each task not being queued until the previous one was completed. It does not guarantee time between two tasks, each new task always gets put on the back of the queue so it can still take a while. If you have concerns about timing then everything needs to be done in the same transaction.

**Comment by MikeOverkamp-diag** [comment]
I wasn't sure on that I I couldn't find a definitive answer in the docs either. In that case I'd suggest just adding the checks to the update task. We should probably move it to the 2xl queue as well.
